### DISCUSSION TITLE
Encounter `TypeError` message occasionally

### DISCUSSION THREAD

**Discussion by tommy-ng0** [serverless/serverless#11280] [Q&A]
Hey all,

We encountered this issue a few times before but we have seen this issue more frequently. We run sls deploy on Gitlab CI/CD:
```
AWS_SDK_LOAD_CONFIG=1   sls deploy --verbose --stage team --region us-west-2 --config containers.yml
Running "serverless" from node_modules
Environment: linux, node 12.22.12, framework 3.21.0 (local) 3.16.0v (global), plugin 6.2.2, SDK 4.3.2
```

and receive the error:
```
TypeError: Cannot read property 'shift' of undefined
    at VariablesResolver.resolveVariable (node_modules/serverless/lib/configuration/variables/resolve.js:230:28)
    at async node_modules/serverless/lib/configuration/variables/resolve.js:118:9
    at async Promise.all (index 0)
    at async VariablesResolver.resolveVariables (node_modules/serverless/lib/configuration/variables/resolve.js:112:5)
    at async VariablesResolver.self (node_modules/serverless/lib/configuration/variables/resolve.js:457:9)
    at async Promise.all (index 0)
    at async VariablesResolver.self (node_modules/serverless/lib/configuration/variables/resolve.js:547:9)
    at async Promise.all (index 6)
    at async node_modules/serverless/scripts/serverless.js:462:11
    at async node_modules/serverless/scripts/serverless.js:237:9
```
but then if we rerun the job, it usually succeeds.

This stack consists of multiple ECS services, tasks and ECS instances. As well as a few EC2 instances. It references ~70 files e.g. `${file(...)}`. Not sure if I have reached any limitation here?

**Comment by pgrzesik** [comment]
Hey @tommysitehost - thanks for letting us know and sorry you've run into trouble. Do you use any plugins in your setup? You mentioned it only happens from time to time, did you manage to find some patterns for when that error is actually raised?

**Reply by tommy-ng0** [reply]
Hi @pgrzesik, thank you for your reply. No we don't use any plugin in this particular stack. So it happened occasionally before (like once or twice in a couple months) so we didn't really look at it. But we added a bunch of Cloudwatch alarms into the `containers.yml` and now it is happening almost every deploy. Our Gitlab Runners are based in Sydney and we see that deploying to US regions encounter this issue a lot more than when we deploy it to Sydney region. However, Sydney region still happens from time to time. A retry would almost always fix this issue but  sometimes we need to run it a few times. I can't see a pattern here except when the yml file is getting larger, it's more likely to hit this error. 

Out of curiosity, what does this error mean?

**Reply by pgrzesik** [reply]
Thanks a lot for additional context, I took a look at it and I'm not sure why it's popping up randomly and why `sources` are not defined in this particular case. The errors likely means that there is a bug in our internal logic, where sources are not populated.

@medikoo Maybe you have an idea about what might be happening here?

**Comment by medikoo** [comment]
This error looks as clear programmer error on Framework side. Let me create bug report from it, and continue over there.
